Use optimized graph in Inline (essentially, making Inline recursive now).

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26703 Remove convert_to_ssa argument from runCleanupPasses - it is only used in one place.
* #26705 Use optimized_graph in graph_executor.
* **#26489 Use optimized graph in Inline (essentially, making Inline recursive now).**
* #26488 Add 'optimized_graph' to Function.
* #26571 Use std::mutex instead of std::call_once in Function when we initialize GraphExecutor.
* #26487 Remove 'recurse' parameter from Inline.

This basically fixes Inline(recurse=true) and makes it a default. One
reservation against running inlining recursively in the original
implementation was that we might hit a quadratic behavior, but in this
implementation it's not an issue since we're inlining only already
inlined graphs and as we recursively descend the call tree we're caching
graphs we've already optimized.

Differential Revision: [D17485744](https://our.internmc.facebook.com/intern/diff/D17485744)

Commit:
Use optimized graph in Inline (essentially, making Inline recursive now).

This basically fixes Inline(recurse=true) and makes it a default. One
reservation against running inlining recursively in the original
implementation was that we might hit a quadratic behavior, but in this
implementation it's not an issue since we're inlining only already
inlined graphs and as we recursively descend the call tree we're caching
graphs we've already optimized.

[ghstack-poisoned]

Commit:
Update on "Use optimized graph in Inline (essentially, making Inline recursive now)."


This basically fixes Inline(recurse=true) and makes it a default. One
reservation against running inlining recursively in the original
implementation was that we might hit a quadratic behavior, but in this
implementation it's not an issue since we're inlining only already
inlined graphs and as we recursively descend the call tree we're caching
graphs we've already optimized.

Differential Revision: [D17485744](https://our.internmc.facebook.com/intern/diff/D17485744)

[ghstack-poisoned]

Commit:
Update on "Use optimized graph in Inline (essentially, making Inline recursive now)."


This basically fixes Inline(recurse=true) and makes it a default. One
reservation against running inlining recursively in the original
implementation was that we might hit a quadratic behavior, but in this
implementation it's not an issue since we're inlining only already
inlined graphs and as we recursively descend the call tree we're caching
graphs we've already optimized.

Differential Revision: [D17485744](https://our.internmc.facebook.com/intern/diff/D17485744)

[ghstack-poisoned]

Commit:
Update on "Use optimized graph in Inline (essentially, making Inline recursive now)."


This basically fixes Inline(recurse=true) and makes it a default. One
reservation against running inlining recursively in the original
implementation was that we might hit a quadratic behavior, but in this
implementation it's not an issue since we're inlining only already
inlined graphs and as we recursively descend the call tree we're caching
graphs we've already optimized.

Differential Revision: [D17485744](https://our.internmc.facebook.com/intern/diff/D17485744)

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good

jerryzh168 commented:
btw, so do we now either don't inline or recursively inline all graphs? Is there a use case to inline one top level graph only like what previous inline API supports?

ZolotukhinM commented:
> btw, so do we now either don't inline or recursively inline all graphs?

Correct.

> Is there a use case to inline one top level graph only like what previous inline API supports?

Not really. It might make sense to not inline everything, but that decision would be made on the entire callgraph anyway (not on a single layer of it). Inlining everything from just one layer doesn't seem to be very useful, especially taking into account arbitrary nature of these layers.

Commit:
Update on "Use optimized graph in Inline (essentially, making Inline recursive now)."


This basically fixes Inline(recurse=true) and makes it a default. One
reservation against running inlining recursively in the original
implementation was that we might hit a quadratic behavior, but in this
implementation it's not an issue since we're inlining only already
inlined graphs and as we recursively descend the call tree we're caching
graphs we've already optimized.

Differential Revision: [D17485744](https://our.internmc.facebook.com/intern/diff/D17485744)

[ghstack-poisoned]
